osmo liquid pooler (#147)
timewave-computer/covenants: contracts/osmo-liquid-pooler/README.md

# osmo liquid pooler

> Contract responsible for providing liquidity to a specified pool on osmosis dex.

Reviewer: ```suggestion
Contract responsible for providing liquidity to a specified pool on the Osmosis dex. Currently we only support GAMM pools where both tokens have equal weights.
```

The contract receives the target denoms, provides liquidity to the specified

> pool, and forwards the LP tokens to the holder.

Reviewer: > and forwards the LP tokens to the holder.

Is this still accurate? Shouldn't the LP tokens remain in the LPer contract?

Works in tandem with the [osmosis liquid pooler outpost](../outpost-osmo-liquid-pooler/README.md),
in order to ensure atomic liquidity provision given the ibc nature of this design.

## flow

The expected state transitions are as follows:

### 1. `Instantiated`

> Ticks incoming to a contract in instantiated state will attempt to create a proxy.

Reviewer: ```suggestion
Ticks incoming to a contract in instantiated state will attempt to create a proxy account on Osmosis via [Polytone](https://github.com/DA0-DA0/polytone)
```

After proxy is created, voice will submit a callback to note. With that callback,
note associates this contract address with the created proxy address and exposes
this association via `RemoteAddress` query. Note also calls back into our contract
which is expecting a callback.

In our contract callback handler, we will query the note for our remote address.
There are two possible cases here:

First case - no address is returned. This means that something went wrong.
We do not advance our state machine and remain in `Instantiated` state.
This means that upon next tick, we will repeat this process until note returns
an address.

Second case - an address is returned. We then store that address, and advance
the state machine to `ProxyCreated`.

### 2. `ProxyCreated`

Ticks incoming to a contract with a created proxy will atempt to fund the proxy.
Proxy being funded is a prerequisite for providing liquidity, and we only want to
attempt providing liquidity if we have delivered all of our funds.

Because of the async nature of IBC, we need to keep things relevant for providing
liquidity up to date. One of such things are balances of our proxy account. Upon

> contract instantiation, we do not store any balances, because we do not even have

Reviewer: ```suggestion
contract instantiation, we do not store balances of the proxy account, because we do not even have
```
